STM32U0: provide support for I2C peripherals

djiatsaf-st opened the pull request:
This PR adds I2C peripherals on the stm32U0 boards.

erwango commented on dts/arm/st/u0/stm32u0.dtsi:
```diff
@@ -207,6 +208,42 @@
+			#size-cells = <0>;
+			reg = <0x40008800 0x400>;
+			clocks = <&rcc STM32_CLOCK_BUS_APB1 0x00800000>;
+			interrupts = <24 0>;
```
Question here.
It seems that instances 2, 3 and 4 will share the same interrupt line. Issue is that there is nothing to handle this case from what I can see at first sight.
Please confirm and see how to address:
- add support using SHARED_IRQ
- generate an error on unsupported configuration
- ...

erwango commented on boards/st/nucleo_u031r8/doc/index.rst:
please update board's .yaml file

erwango commented on dts/bindings/i2c/st,stm32u0-i2c.yaml:
This commit should come first to allow compilation of others.

erwango commented on dts/bindings/i2c/st,stm32u0-i2c.yaml:
```diff
@@ -0,0 +1,13 @@
+        STM32U0 I2C Controller
+        This compatibility applies to all I2C hardware blocks corresponding
+        to the version available in STM32U0 SoCs.
+        This version of the STM32 I2C hardware block can be identified by the
+        absence of the SMBHEN and SMBDEN bits in the I2C_CR1 register.
+        For example, but not limited to: STM32U0
```
Could be summarized by "[...] doesn't support SMBUS"

ydamigos commented on drivers/i2c/i2c_ll_stm32.c:
```diff
@@ -115,7 +115,9 @@ int i2c_stm32_runtime_configure(const struct device *dev, uint32_t config)
 #endif
 
 	LL_I2C_Disable(i2c);
+#if !DT_HAS_COMPAT_STATUS_OKAY(st_stm32u0_i2c)
```
Are we going to follow this approach or check if the specific bits are defined in SoC's header file (e.g. `I2C_CR1_SMBDEN`) like we do elsewhere in I2C driver (e.g. `I2C_FLTR_ANOFF`)?

djiatsaf-st replied:
Hello @ydamigos ,
Thanks for the suggestion, I think I'll do it this way, it works and I don't have any other constraints with it.

ydamigos replied:
I would avoid to introduce a new compatible `st,stm32u0-i2c`, just to exclude code, while there is another way doing it which is already there and it is used in STM32 I2C driver. But @erwango has the last word as the maintainer.

erwango replied:
Thanks @ydamigos for the proposal. I agree that for now, a new compatible doesn't add much value and is more complex. Let's go with the register presence check for now, we'll still be able to come back on that later if required. Thanks for the proposal, @ydamigos!

djiatsaf-st replied:
> I would avoid to introduce a new compatible `st,stm32u0-i2c`, just to exclude code, while there is another way doing it which is already there and it is used in STM32 I2C driver. But @erwango has the last word as the maintainer.

hello @ydamigos 
I think my last post was confusing, I said I tested your suggestion and it worked, so I would use that instead , it's simpler indeed.

djiatsaf-st committed:
soc: st: stm32: stm32u0x: add soc configs for i2c shared irq

check if two or three I2C instances with same irq are enabled
at same time then enable shared_interrupt handler.

Signed-off-by: Fabrice DJIATSA <fabrice.djiatsa-ext@st.com>

djiatsaf-st committed:
dts: arm: st: u0: add i2c nodes in dtsi files

we have four i2C peripherals .
- three shared between stm32u031/73/83
- One between stm32u073/83

Signed-off-by: Fabrice DJIATSA <fabrice.djiatsa-ext@st.com>

djiatsaf-st committed:
boards: st: add i2c nodes in dts and update docs

-enable i2c in nucleo_u031r8, nucleo_u083rc and stm32u083c_dk
- update documentations

Signed-off-by: Fabrice DJIATSA <fabrice.djiatsa-ext@st.com>

djiatsaf-st committed:
drivers: i2c: update i2c driver to integrate stm32u0

- enabled combined interrupt configuration for stm32u0x serie.
- since stm32u0 serie doesn't support SMBUS we need to avoid
use SMBHEN and SMBDEN bits register.

Signed-off-by: Fabrice DJIATSA <fabrice.djiatsa-ext@st.com>

djiatsaf-st committed:
tests: drivers: i2c: i2c_target_api: update i2c_target_api

- add overlays and kconfigs files for nucleo_u083rc and
stm32u083c_dk boards.
- update testcase.yaml for CI

Signed-off-by: Fabrice DJIATSA <fabrice.djiatsa-ext@st.com>

erwango commented on soc/st/stm32/stm32u0x/Kconfig.defconfig.stm32u073xx:
```diff
@@ -8,4 +8,24 @@ if SOC_STM32U073XX
 
+if I2C_STM32
+
+config I2C2_I2C3_ENABLED
+	default y if $(dt_nodelabel_enabled,i2c2) && $(dt_nodelabel_enabled,i2c3)
+
+config I2C2_I2C4_ENABLED
+	default y if $(dt_nodelabel_enabled,i2c2) && $(dt_nodelabel_enabled,i2c4)
+
+config I2C3_I2C4_ENABLED
+	default y if $(dt_nodelabel_enabled,i2c3) && $(dt_nodelabel_enabled,i2c4)
+
+config I2C2_I2C3_I2C4_ENABLED
+	default y if $(dt_nodelabel_enabled,i2c2) && $(dt_nodelabel_enabled,i2c3) &&  $(dt_nodelabel_enabled,i2c4)
+
+
+config SHARED_INTERRUPTS
+	default y if I2C2_I2C3_ENABLED || I2C3_I2C4_ENABLED || I2C2_I2C4_ENABLED || I2C2_I2C3_I2C4_ENABLED
```
Could be summed up as:
```suggestion
config SHARED_INTERRUPTS
	default y if $(dt_nodelabel_enabled,i2c2) && $(dt_nodelabel_enabled,i2c3)
	default y if $(dt_nodelabel_enabled,i2c2) && $(dt_nodelabel_enabled,i2c4)
	default y if $(dt_nodelabel_enabled,i2c3) && $(dt_nodelabel_enabled,i2c4)
```

and  can be put in `soc/st/stm32/stm32u0x/Kconfig.defconfig`

ydamigos commented on drivers/i2c/i2c_ll_stm32.c:
```diff
@@ -115,7 +115,9 @@ int i2c_stm32_runtime_configure(const struct device *dev, uint32_t config)
 #endif
 
 	LL_I2C_Disable(i2c);
+#if defined(SMBDEN) && defined(SMBHEN)
```
```suggestion
#if defined(I2C_CR1_SMBHEN) && defined(I2C_CR1_SMBHEN)
```

ydamigos reviewed (approved):
LGTM
